File: cpp_quant/src/backtest.cpp
#include "backtest.h"
#include <cmath>
#include <algorithm>

std::tuple<double, double, double, double, int> 
run_ml_backtest_cpp(
    const std::vector<double>& prices,
    const std::vector<double>& probabilities,
    double threshold,
    double initial_capital,
    double commission,
    double slippage
) {
    double capital = initial_capital;
    double position = 0.0;
    double entry_price = 0.0;
    int trades = 0;
    int wins = 0;
    std::vector<double> equity_curve;
    equity_curve.reserve(prices.size());
    equity_curve.push_back(initial_capital);

    size_t n = std::min(prices.size(), probabilities.size());

    for (size_t i = 0; i < n; ++i) {
        double price = prices[i];
        double prob = probabilities[i];

        if (prob > threshold && position == 0.0) {
            double cost_per_unit = price * (1.0 + commission + slippage);
            double size = capital / cost_per_unit;
            if (size > 0.0) {
                capital -= size * cost_per_unit;
                position = size;
                entry_price = price;
            }
        }
        else if (prob < (1.0 - threshold) && position > 0.0) {
            double revenue_per_unit = price * (1.0 - commission - slippage);
            double revenue = position * revenue_per_unit;
            double pnl = revenue - position * entry_price;
            capital += revenue;
            if (pnl > 0) ++wins;
            ++trades;
            position = 0.0;
            entry_price = 0.0;
        }
        double equity = capital + position * price;
        equity_curve.push_back(equity);
    }

    if (position > 0.0 && !prices.empty()) {
        double closing_price = prices.back();
        double revenue_per_unit = closing_price * (1.0 - commission - slippage);
        double revenue = position * revenue_per_unit;
        double pnl = revenue - position * entry_price;
        capital += revenue;
        if (pnl > 0) ++wins;
        ++trades;
    }

    double total_return = (capital - initial_capital) / initial_capital;

    double sharpe = 0.0;
    if (equity_curve.size() > 1) {
        std::vector<double> returns;
        returns.reserve(equity_curve.size()-1);
        for (size_t i = 1; i < equity_curve.size(); ++i) {
            double ret = (equity_curve[i] - equity_curve[i-1]) / equity_curve[i-1];
            returns.push_back(ret);
        }
        double mean = 0.0;
        for (double r : returns) mean += r;
        mean /= returns.size();
        double stddev = 0.0;
        for (double r : returns) stddev += (r - mean) * (r - mean);
        stddev = std::sqrt(stddev / returns.size());
        if (stddev > 1e-8) {
            sharpe = (mean / stddev) * std::sqrt(252.0);
        }
    }

    double max_drawdown = 0.0;
    double peak = equity_curve[0];
    for (double e : equity_curve) {
        if (e > peak) peak = e;
        double dd = (peak - e) / peak;
        if (dd > max_drawdown) max_drawdown = dd;
    }
    max_drawdown = -max_drawdown;

    double win_rate = (trades > 0) ? static_cast<double>(wins) / trades : 0.0;

    return {total_return, sharpe, max_drawdown, win_rate, trades};
}

File: researchos/quant_engine/backtest.py
"""
Backtest engine with proper trade recording.
"""

from dataclasses import dataclass
from typing import Any

import numpy as np


@dataclass
class BacktestResult:
    total_return: float
    sharpe_ratio: float
    max_drawdown: float
    win_rate: float
    num_trades: int
    signals: list[Any]


class BacktestEngine:
    def __init__(self, initial_capital=100000.0, commission=0.001, slippage=0.0005):
        self.initial_capital = initial_capital
        self.commission = commission
        self.slippage = slippage

    def run(self, prices: list[float], strategy) -> BacktestResult:
        signals = strategy.generate_signals(prices)
        if not signals:
            return BacktestResult(0.0, 0.0, 0.0, 0.0, 0, signals)

        capital = self.initial_capital
        position = 0.0
        entry_price = 0.0
        trades = []  # (action, price, size, pnl)
        equity_curve = [capital]

        for signal in signals:
            price = signal.price
            if signal.action == "BUY" and position == 0:
                cost_per_unit = price * (1 + self.commission + self.slippage)
                size = capital / cost_per_unit
                if size > 0:
                    capital -= size * cost_per_unit
                    position = size
                    entry_price = price
                    trades.append(("BUY", price, size, 0.0))

            elif signal.action == "SELL" and position > 0:
                revenue_per_unit = price * (1 - self.commission - self.slippage)
                revenue = position * revenue_per_unit
                pnl = revenue - position * entry_price
                capital += revenue
                trades.append(("SELL", price, position, pnl))
                position = 0.0
                entry_price = 0.0

            # Хөрөнгийн үнэлгээ
            current_equity = capital + position * price
            equity_curve.append(current_equity)

        # Хэрэв позиц үлдсэн бол эцсийн үнээр хаах
        if position > 0 and prices:
            closing_price = prices[-1]
            revenue_per_unit = closing_price * (1 - self.commission - self.slippage)
            revenue = position * revenue_per_unit
            pnl = revenue - position * entry_price
            capital += revenue
            trades.append(("CLOSE", closing_price, position, pnl))
            position = 0.0

        final_value = capital
        total_return = (final_value - self.initial_capital) / self.initial_capital

        # Sharpe ratio
        equity = np.array(equity_curve)
        returns = np.diff(equity) / equity[:-1]
        if len(returns) > 1:
            sharpe = np.mean(returns) / (np.std(returns) + 1e-8) * np.sqrt(252)
        else:
            sharpe = 0.0

        # Max drawdown
        peak = np.maximum.accumulate(equity)
        drawdown = (peak - equity) / peak
        max_drawdown = -np.max(drawdown) if len(drawdown) > 0 else 0.0

        # Win rate – зөвхөн хаагдсан трейдууд (SELL эсвэл CLOSE)
        closed_trades = [t for t in trades if t[0] in ("SELL", "CLOSE")]
        winning_trades = [t for t in closed_trades if t[3] > 0]  # t[3] = pnl
        win_rate = len(winning_trades) / len(closed_trades) if closed_trades else 0.0

        return BacktestResult(
            total_return=total_return,
            sharpe_ratio=sharpe,
            max_drawdown=max_drawdown,
            win_rate=win_rate,
            num_trades=len(closed_trades),
            signals=signals,
        )
